### PR TITLE
YL #2 save image link

### LINKED ISSUE #2: Save Image Link
#### As a user I want to save a link to an image.

__Story__:

Now that we have an awesome landing page for our users to experience it is time
to introduce our first feature! In general, we want to deliver value to our
customers as fast as possible. As a result, we have decided that we want our
users to be able to save links to images, and immediately see that we have
saved the URL to the image.

One could argue that simply saving the image, and not showing it, provides some
value. Yes, you could split this story here. However, because saving image URLs
and viewing those images are both relatively small additions that work well
together, it is acceptable to complete them together. Note, on the other hand,
that we do not yet want to implement the image index. Its addition adds too
much complexity, so we’ll save that for a future story.

In general, feature addition, or stories, should be split into the smallest
deliverable chunk that provides value to your users with the caveat that it
might make sense to combine two, sometimes more, tiny and related features into
a single story.

__Acceptance criteria__:
- [x] The landing page links to the image link submission form.
- [x] The link is entered through a form.
- [x] After the form is saved, the link is persisted in the database.
- [x] After the form is submitted, the user is redirected to a page displaying
  the image.
- [x] I cannot successfully save an image with an invalid URL.
- [x] An error message is associated with the appropriate input field on
  failure.

__Discussion Topic__:
- [ ] What is a valid image URL?

__Reference material__:
- https://github.com/plataformatec/simple_form

__Dependencies__:
- Deployment Environment

__Note__: Many image sharing sites will permit one to upload images, or
download images when provided a link. For this project we will only work with
the link, i.e., URL, to various images.

__Note__: To understand the database tables created by the `acts_as_taggable` gem, you may find it useful to use a program like the [DB Browser for SQLite](http://sqlitebrowser.org/).

### PR DESCRIPTION
**Pull request by YJ83Lee**
Closes #2 

---

The following is a checklist to remind you of the steps to take when opening a
new PR. The horizontal line above and everything below can be removed from the
pull request description.

- [x] I have replaced `X` above with the associated issue number.
- [x] I have changed the issue's label to "pull request".
- [x] I have assigned the designated reviewer to this PR.
- [x] I have set the appropriate label on this PR.
- [x] I have written clear commit messages ([ref](http://chris.beams.io/posts/git-commit/)).

If needed, please review the document on the
[Image Sharing Project](https://sites.google.com/a/appfolio.com/eng/new-hire-resources/engineering-academy/ropes/image-sharing-project)
for information on how labels should be used.

### CODE REVIEW

File: app/controllers/imagelinks_controller.rb [hunk 1]
```diff
@@ -1,5 +1,7 @@
 class ImagelinksController < ApplicationController
-  def index; end
+  def index
```
[the remark this answers is not in the document]
Comment (YJ83Lee) [reply app/controllers/imagelinks_controller.rb]: I rechecked https://github.com/AppFolioOnboarding/image-sharer-YJ83Lee/issues/2 and figured that it did mentioned

> After the form is submitted, the user is redirected to a page displaying
  the image.

in the Acceptance criteria.

File: app/controllers/imagelinks_controller.rb [hunk 2]
```diff
@@ -4,4 +4,13 @@ def index; end
   def new
     @imagelink = Imagelink.new
   end
+
+  def create
+    @new_imagelink = Imagelink.new link: params[:imagelink][:link]
+    if @new_imagelink.save
+      redirect_to action: 'index'
+    else
+      redirect_to action: 'new', alert: "#{params[:imagelink][:link]} is an invalid URL"
```
[the remark this answers is not in the document]
Comment (YJ83Lee) [reply app/controllers/imagelinks_controller.rb]: The way it currently works is that it adds a query string like `'?alert=invalidURL'`.
I believe flash messages (cookie based) is better, will try it.

File: app/views/imagelinks/new.erb [hunk 3]
```diff
@@ -1,4 +1,7 @@
 <h1>Create new image link</h1>
+<% if params[:alert] %>
```
[the remark this answers is not in the document]
Comment (YJ83Lee) [reply app/views/imagelinks/new.erb]: I have not started using layout yet.
Will consider moving it once layout is used.

File: config/routes.rb [hunk 4]
```diff
@@ -1,7 +1,7 @@
 Rails.application.routes.draw do
   # For details on the DSL available within this file, see http://guides.rubyonrails.org/routing.html
 
-  root 'imagelinks#index'
+  root 'imagelinks#new'
 
   resources :imagelinks
```
[the remark this answers is not in the document]
Comment (YJ83Lee) [reply config/routes.rb]: I totally agree. Will fix it.

File: app/controllers/imagelinks_controller.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (YJ83Lee) [reply app/controllers/imagelinks_controller.rb]: Fixed in https://github.com/AppFolioOnboarding/image-sharer-YJ83Lee/pull/16/commits/aa4fc224f815de8c4ca961059f03337635c120c1

File: app/controllers/imagelinks_controller.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (YJ83Lee) [reply app/controllers/imagelinks_controller.rb]: Showing actual images in https://github.com/AppFolioOnboarding/image-sharer-YJ83Lee/pull/16/commits/2a8a9b96500cabdc89c215b9d3feba9a10e74ddb.

File: config/routes.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (YJ83Lee) [reply config/routes.rb]: Fixed by https://github.com/AppFolioOnboarding/image-sharer-YJ83Lee/pull/16/commits/1cc4f497402bcf2625da1e7ff1ded5a0b2d532a4.

File: test/fixtures/imagelinks.yml [hunk 5]
```diff
@@ -0,0 +1,7 @@
+# Read about fixtures at http://api.rubyonrails.org/classes/ActiveRecord/FixtureSet.html
+
+one:
```
[the remark this answers is not in the document]
Comment (YJ83Lee) [reply test/fixtures/imagelinks.yml]: Removed by https://github.com/AppFolioOnboarding/image-sharer-YJ83Lee/pull/16/commits/7a9e31700a63cb8f802fc2774ec4377bb594327a.

File: test/models/imagelink_test.rb [hunk 6]
```diff
@@ -0,0 +1,8 @@
+require 'test_helper'
+
+class ImagelinkTest < ActiveSupport::TestCase
+  test 'should not save imagelink without a link' do
+    imagelink = Imagelink.new
+    assert_not imagelink.save
```
[the remark this answers is not in the document]
Comment (YJ83Lee) [reply test/models/imagelink_test.rb]: Fixed in https://github.com/AppFolioOnboarding/image-sharer-YJ83Lee/pull/16/commits/c19379239fabee52216ae1c3cb0249ed3e75a39c.

File: app/models/imagelink.rb [hunk 7]
```diff
@@ -1,3 +1,4 @@
 class Imagelink < ApplicationRecord
   validates :link, presence: true
+  validates :link, format: URI.regexp(%w[http https])
```
[the remark this answers is not in the document]
Comment (YJ83Lee) [reply app/models/imagelink.rb]: Introduced validates_url in https://github.com/AppFolioOnboarding/image-sharer-YJ83Lee/pull/16/commits/c19379239fabee52216ae1c3cb0249ed3e75a39c.

File: app/controllers/imagelinks_controller.rb [hunk 8]
```diff
@@ -4,4 +4,13 @@ def index; end
   def new
     @imagelink = Imagelink.new
   end
+
+  def create
+    @new_imagelink = Imagelink.new link: params[:imagelink][:link]
+    if @new_imagelink.save
+      redirect_to action: 'index'
+    else
```
Comment (almgong) [inline app/controllers/imagelinks_controller.rb]: Should we add test coverage for this case where the `ImageLink` is not valid? I.e when the image link is not a valid URL?

Comment (almgong) [reply app/controllers/imagelinks_controller.rb]: `assert_select`

Comment (YJ83Lee) [reply app/controllers/imagelinks_controller.rb]: Fixed in https://github.com/AppFolioOnboarding/image-sharer-YJ83Lee/pull/16/commits/be140f1764af99381371099884c1e17a76b3d45f.

**Pull request merged**